File: cpp/sudoku.cpp
#include <iostream>
#include <string>
#include <cmath>

bool isSolved = false;
int sudoku[9][9] = {
    {0, 0, 0, 0, 0, 9, 8, 0, 0},
    {0, 0, 0, 0, 7, 0, 0, 6, 2},
    {8, 0, 0, 0, 0, 2, 0, 0, 7},

    {0, 0, 7, 8, 0, 0, 4, 0, 0},
    {2, 0, 0, 0, 3, 6, 0, 0, 0},
    {5, 0, 0, 0, 0, 0, 0, 0, 0},

    {0, 0, 0, 0, 0, 0, 0, 5, 0},
    {0, 8, 9, 1, 0, 0, 0, 0, 6},
    {0, 0, 0, 4, 0, 0, 7, 0, 0}};

bool CheckValid(int i, int j, int v);
bool Solve();

bool Solve()
{
    for (int i = 0; i < 9; i++)
    {
        for (int j = 0; j < 9; j++)
        {
            if (sudoku[i][j] == 0)
            {
                for (int v = 1; v <= 9; v++)
                {
                    if (CheckValid(i, j, v))
                    {
                        sudoku[i][j] = v;
                        Solve();
                        if (isSolved)
                        {
                            return true;
                        }
                        sudoku[i][j] = 0;
                    }
                }
                return true;
            }
        }
    }
    isSolved = true;
    return true;
}

bool CheckValid(int i, int j, int v)
{
    for (int x = 0; x < 9; x++)
    {
        if (i == x) continue;
        if (sudoku[x][j] == v) return false;
    }
    for (int y = 0; y < 9; y++)
    {
        if (j == y) continue;
        if (sudoku[i][y] == v) return false;
    }
    int xBegin = (int)floor((double)i / 3) * 3;
    int yBegin = (int)floor((double)j / 3) * 3;

    for (int x = xBegin; x < xBegin + 3; x++)
    {
        for (int y = yBegin; y < yBegin + 3; y++)
        {
            if (i == x && j == y) continue;
            if (sudoku[x][y] == v) return false;
        }
    }
    return true;
}

void Display(int array[9][9])
{
    std::cout << "+-------+-------+-------+" << std::endl;

    for (int i = 0; i < 9; i++)
    {
        std::cout << "| ";
        for (int j = 0; j < 9; j++)
        {
            if (array[i][j] == 0)
            {
                std::cout << "  ";
            }
            else
            {
                std::cout << array[i][j] << " ";
            }

            if ((j + 1) % 3 == 0)
            {
                std::cout << "| ";
            }
        }

        if ((i + 1) % 3 == 0)
        {
            std::cout << std::endl
                      << "+-------+-------+-------+";
        }
        std::cout << std::endl;
    }
}

int main()
{
    Display(sudoku);
    Solve();
    Display(sudoku);
    return 0;
}

File: julia/sudoku.jl
isSolved = false
sudoku =[
    [0, 0, 0, 0, 0, 9, 8, 0, 0],
    [0, 0, 0, 0, 7, 0, 0, 6, 2],
    [8, 0, 0, 0, 0, 2, 0, 0, 7],
  
    [0, 0, 7, 8, 0, 0, 4, 0, 0],
    [2, 0, 0, 0, 3, 6, 0, 0, 0],
    [5, 0, 0, 0, 0, 0, 0, 0, 0],
  
    [0, 0, 0, 0, 0, 0, 0, 5, 0],
    [0, 8, 9, 1, 0, 0, 0, 0, 6],
    [0, 0, 0, 4, 0, 0, 7, 0, 0]
  ];

function Solve()
    for i in 0:8
        for j in 0:8
            if (sudoku[i + 1][j + 1] == 0)
                for v in 1:9
                    if (CheckValid(i, j, v))
                        sudoku[i + 1][j + 1] = v;
                        Solve();
                        if( isSolved )
                            return true;
                        end;
                        sudoku[i + 1][j + 1] = 0;
                    end;
                end;
                return true;
            end;
        end;
    end;
    isSolved = true;
    return true;
end;

function CheckValid(i, j, v)
    for x in 0:8
        if (i == x)
            continue;
        end;
        if (sudoku[x + 1][j + 1] == v)
            return false;
        end;
    end;
    for y in 0:8
        if (j == y)
            continue;
        end;
        if (sudoku[i + 1][y + 1] == v)
            return false;
        end;
    end;

    xBegin = Int.(floor(i / 3)) * 3;
    yBegin = Int.(floor(j / 3)) * 3;

    for x in xBegin:xBegin + 2
        for y in yBegin:yBegin + 2
            if (i == x && j == y)
                continue;
            end;
            if (sudoku[x + 1][y + 1] == v)
                return false;
            end;
        end;
    end;

    return true;
end;

function Display()
    print("+-------+-------+-------+\r\n");
    for i in 1:9
        print("| ");
        for j in 1:9
            if( sudoku[i][j] == 0)
                print("  ");
            else
                print(sudoku[i][j]);
                print(" ");
            end
            if (j % 3 == 0)
                print("| ");
            end
        end;
        if (i % 3 == 0)
            print("\r\n+-------+-------+-------+");
        end
        print("\r\n");
    end;
end;

Display();
Solve();
Display();